File: packages/downsample/typescript/src/gaussian-kernel-radius.ts
import {
  InterfaceTypes,
  PipelineOutput,
  PipelineInput,
  runPipeline
} from 'itk-wasm'

import GaussianKernelRadiusOptions from './gaussian-kernel-radius-options.js'
import GaussianKernelRadiusResult from './gaussian-kernel-radius-result.js'

import { getPipelinesBaseUrl } from './pipelines-base-url.js'
import { getPipelineWorkerUrl } from './pipeline-worker-url.js'

import { getDefaultWebWorker } from './default-web-worker.js'

/**
 * Radius in pixels required for effective discrete gaussian filtering.
 *
 * @param {GaussianKernelRadiusOptions} options - options object
 *
 * @returns {Promise<GaussianKernelRadiusResult>} - result object
 */
async function gaussianKernelRadius(
  options: GaussianKernelRadiusOptions = { size: [1,1], sigma: [1,1], }
) : Promise<GaussianKernelRadiusResult> {

  const desiredOutputs: Array<PipelineOutput> = [
    { type: InterfaceTypes.JsonCompatible },
  ]

  const inputs: Array<PipelineInput> = [
  ]

  const args = []
  // Inputs
  // Outputs
  const radiusName = '0'
  args.push(radiusName)

  // Options
  args.push('--memory-io')
  if (typeof options.size !== "undefined") {
    if(options.size.length < 1) {
      throw new Error('"size" option must have a length > 1')
    }
    args.push('--size')

    await Promise.all(options.size.map(async (value) => {
      args.push(value.toString())

    }))
  }
  if (typeof options.sigma !== "undefined") {
    if(options.sigma.length < 1) {
      throw new Error('"sigma" option must have a length > 1')
    }
    args.push('--sigma')

    await Promise.all(options.sigma.map(async (value) => {
      args.push(value.toString())

    }))
  }
  if (typeof options.maxKernelWidth !== "undefined") {
    args.push('--max-kernel-width', options.maxKernelWidth.toString())

  }
  if (typeof options.maxKernelError !== "undefined") {
    args.push('--max-kernel-error', options.maxKernelError.toString())

  }

  const pipelinePath = 'gaussian-kernel-radius'

  let workerToUse = options?.webWorker
  if (workerToUse === undefined) {
    workerToUse = await getDefaultWebWorker()
  }
  const {
    webWorker: usedWebWorker,
    returnValue,
    stderr,
    outputs
  } = await runPipeline(pipelinePath, args, desiredOutputs, inputs, { pipelineBaseUrl: getPipelinesBaseUrl(), pipelineWorkerUrl: getPipelineWorkerUrl(), webWorker: workerToUse, noCopy: options?.noCopy })
  if (returnValue !== 0 && stderr !== "") {
    throw new Error(stderr)
  }

  const result = {
    webWorker: usedWebWorker as Worker,
    radius: outputs[0]?.data as number[],
  }
  return result
}

export default gaussianKernelRadius

File: packages/downsample/gaussian-kernel-radius.cxx
/*=========================================================================

 *  Copyright NumFOCUS
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/
#include "itkPipeline.h"
#include "itkInputTextStream.h"
#include "itkOutputTextStream.h"

#include "itkGaussianOperator.h"

#include "rapidjson/document.h"
#include "rapidjson/stringbuffer.h"
#include "rapidjson/writer.h"


int main(int argc, char * argv[])
{
  itk::wasm::Pipeline pipeline("gaussian-kernel-radius", "Radius in pixels required for effective discrete gaussian filtering.", argc, argv);

  std::vector<uint64_t> size { 1, 1 };
  pipeline.add_option("--size", size, "Size in pixels")->required();

  std::vector<double> sigma { 1.0, 1.0 };
  pipeline.add_option("--sigma", sigma, "Sigma in pixel units")->required();

  uint64_t maxKernelWidth = 32;
  pipeline.add_option("--max-kernel-width", maxKernelWidth, "Maximum kernel width in pixels.");

  double maxKernelError = 0.01;
  pipeline.add_option("--max-kernel-error", maxKernelError, "Maximum kernel error.");

  itk::wasm::OutputTextStream radiusStream;
  pipeline.add_option("radius", radiusStream, "Output kernel radius.")->required()->type_name("OUTPUT_JSON");

  ITK_WASM_PARSE(pipeline);

  if (size.size() != sigma.size())
  {
    std::ostringstream ostrm;
    ostrm << "Size and sigma must have the same length. Size: " << size.size() << " Sigma: " << sigma.size() << ".\n";
    CLI::Error err("Runtime error", ostrm.str(), 1);
    return pipeline.exit(err);
  }

  rapidjson::Document radiusArray;
  radiusArray.SetArray();
  rapidjson::Document::AllocatorType& allocator = radiusArray.GetAllocator();
  for (size_t i = 0; i < size.size(); ++i)
  {
    auto gaussianOperator = itk::GaussianOperator<double, 1>();
    gaussianOperator.SetDirection(0);
    // at most the size in the same direction of a chunk
    const auto constrainedMaxKernelWidth = std::min(maxKernelWidth, size[i]);
    gaussianOperator.SetMaximumKernelWidth(constrainedMaxKernelWidth);
    gaussianOperator.SetMaximumError(maxKernelError);
    const double variance = sigma[i] * sigma[i];
    gaussianOperator.SetVariance(variance);
    gaussianOperator.CreateDirectional();
    radiusArray.PushBack(rapidjson::Value().SetInt(gaussianOperator.GetRadius(0)), allocator);
  }

  rapidjson::StringBuffer stringBuffer;
  rapidjson::Writer<rapidjson::StringBuffer> writer(stringBuffer);
  radiusArray.Accept(writer);

  radiusStream.Get() << stringBuffer.GetString();

  return EXIT_SUCCESS;
}

